Made PWM frequency changable by $33 setting

cprezzi opened the pull request:
I have made the PWM frequency changable through the $33 param, so laser engraver users can easily experiment with that param. 

Default frequency is set to 5'000Hz (=200us PWM period), which is a good mix for grayscale engraving and cutting.

Higher frequencies (like 20'000Hz) tend to cut deeper with less burning and lower frequencies (like 2'500Hz) could deliver slightliy better grayscale representation. The effect depends on many factors like the laser power supply, material, focus, feed...

cprezzi committed:
Config for K40

cprezzi committed:
Added $33 for PWM frequency and defaults for K40

cprezzi committed:
Error corrected in config.h (missing ")" on LIMIT_MASK)

cprezzi committed:
Corrected release link in README.md

tbfleming commented:
@chamnit Are you OK reserving $33 for PWM frequency?

@cprezzi @chamnit If the frequency is configurable, then we should probably make the other parameters configurable also. We'd need to reserve config numbers for replacements of:

* SPINDLE_PWM_PERIOD ($33: frequency)
* SPINDLE_PWM_OFF_VALUE ($34?: would be a percentage)
* SPINDLE_PWM_MIN_VALUE ($35?: would be a percentage)
* SPINDLE_PWM_MAX_VALUE ($36?: would be a percentage)

These, together with the existing rpm_min and rpm_max settings would allow full PWM tuning without recompiling.

tbfleming commented:
It's annoying that different K40's have different steps-per-mm.

chamnit commented:
@tbfleming : I'll follow your lead on this one. $32+ settings were intended to be reserved for spindle/laser settings anyway. I'm ok with everything proposed so far, except PWM period should probably be expressed in frequency. The values are easier to understand as Hz, rather than micro- or milli-seconds. 

I was also thinking about making PWM frequency something that can be altered in g-code in the future as an override. This would make it possible for LW to tailor tool paths for different operations that require different frequencies and let users quickly test materials. Thinking that this would be a scalar of the PWM frequency setting like the other overrides.

tbfleming commented:
> except PWM period should probably be expressed in frequency

That's what @cprezzi did :)

> I was also thinking about making PWM frequency something that can be altered in g-code in the future as an override

That would rock!

cprezzi commented:
@tbfleming If we are thinking about additional $settings, what about the PWM pin? This could cancel the need for multiple compiled versions.

cprezzi committed:
Added 4.axes (A) and moved default settings to defaults.h and cpu-map.h

chamnit commented:
@cprezzi : FYI, I'm planning on making just about everything I can a setting. However, I'm struggling to understand why the PWM pin would need one and how it could change.

tbfleming commented:
@chamnit Smoothieboard provides a set of pins to choose from, so different users end up hooking their laser control to different pins. I've been providing .bin files for the two most common choices.

cprezzi commented:
@chamnit There are multiple "smoothie compatible" (LPC1769) boards out there, but not all use the same pin for the Spindle/Laser PWM. It's also depending if the user likes to use the heatbed mosfet output (like for K40 lasers), or a logic level pin (like for diode laser engravers). Mosfet is additionally inverted (pull down).

cprezzi commented:
@chamnit For your info: Today I have implemented a 4. axes (A) into my fork (see https://github.com/cprezzi/grbl-LPC/tree/more-axis). Does it make sence that I do a PR (just for grbl_LPC), or do you plan to add more axis anyway?

cprezzi commented:
@tbfleming In that branch, I have also moved all the default settings and pin mappings to cpu_map.h and defaults.h, so the board and machine can be set by two simple #define in config.h ;)

chamnit commented:
@cprezzi @tbfleming : Ok. Makes sense. However, this is something board specific, rather than a generalized setting. I think it'll create some problems with some of the assumptions I made with the new Grbl HAL. There is no separate set of tools to create these types of settings and make sure they get initialized. Not a big deal, but something I should get installed before public release.

Perhaps we need to declare a set of `$$` setting values that are vendor/board specific. Let's arbitrarily set them to either $50-$99 or greater than some large number like $500 ($100-$250 are reserved for axes settings). This way the new Grbl HAL will be able to simply check the values and send it over a HAL function to deal with them.

chamnit commented:
@cprezzi : I'll add a 4th axis in the next major release. But, I can't say when that will be, because I have a upcoming NASA project. I'm developing a new Mars solar array farm concept pretty much solo. It's going to eat up a huge chunk of my time and energy for the rest of the year, starting in June. However,  I'll find the time somehow, as I always do. So, feel free to do what you need to do. I can always use your code or pull from it when it comes time.

cprezzi commented:
@chamnit I always knew you are the guy for rocket science ;) Good luck! 👍

tbfleming commented:
I added $33-$36 for spindle control. My laser isn't set up right now and I had to return the borrowed oscope, so I can't test it. I pointed users to @cprezzi's branch for releases since that branch is moving faster.

chamnit commented:
Would it help to make @cprezzi a collaborator for this branch?

tbfleming commented:
@chamnit yes please

cprezzi commented:
Thanks @chamnit @tbfleming !

cprezzi committed:
Merge remote-tracking branch 'upstream/master'

cprezzi committed:
Added default settings for PWM_OFF_VALUE, PWM_MIN_VALUE and PMW_MAX_VALUE

cprezzi committed:
Merged new $ params

cprezzi committed:
Merge branch 'master' into more-axis

cprezzi committed:
Corrected stepper invert mask for K40

cprezzi committed:
Corrected K40 defaults

cprezzi committed:
Moved homing origin to bottom/left instead of top/right

cprezzi committed:
Merge branch 'master' into more-axis

cprezzi committed:
Adjusted config.h

cprezzi committed:
Added Links for Make and ARM Toolchain.

cprezzi committed:
Added Links for Make Tool and ARM Toolchain.

cprezzi committed:
Merge branch 'master' of https://github.com/cprezzi/grbl-LPC

cprezzi committed:
Highlight make as command.

cprezzi committed:
Made CPU-MAP per Board instead of MCU

cprezzi committed:
Made A-Axis switchable in config.h

cprezzi committed:
Made board switchable in config.h

cprezzi committed:
Always use MIN endstop pins (independant of board!)

cprezzi committed:
Support coolant flood and mist for remix board

cprezzi committed:
Merge branch 'more-axis'

cprezzi committed:
Set coolant pins for all supported boards in cpu-map.h

cprezzi committed:
Activated HOMING_FORCE_POSITIVE_SPACE and added SPINDLE_PWM_PIN_2_4 param

cprezzi committed:
Corrected G10 L20 P0 command (setZero)

jorgerobles committed:
Added Dockerfile

cprezzi committed:
Merge pull request #6 from jorgerobles/master

Added Dockerfile

cprezzi committed:
Ported spindle enable pin + FEED_HOLD, CYCLE_START and DOOR_SWITCH pins

00alkskodi00 commented:
Hi how can i compile firmware? Thi is for me not working https://github.com/gnea/grbl-LPC/wiki

im trying 5 days but whitout finish...

cprezzi committed:
Added config option to have PWM on P1.23 for MKS SBase

bkp23 committed:
Added a description to each option, including the option codes

cprezzi committed:
Merge pull request #14 from bkp23/defaults_comments

Added a description to each option, including the option codes (ordered by $key).

bkp23 committed:
Added comments throughout code

bkp23 committed:
Revert simple code changes so only comments have been changed

cprezzi committed:
Merge pull request #21 from bkp23/master

Thanks for the helpful code comments!

tscofield committed:
impliment changes to allow probing to work and map the probe pin to p1.27 on the sbase board

tscofield committed:
Add additional info to help new users

cprezzi committed:
Merge pull request #28 from tscofield/build_doc

Add additional docker info to help new users

cprezzi committed:
Merge pull request #27 from tscofield/probe

changes to allow probing on pin 1.27

cprezzi committed:
Merge branch 'master' of https://github.com/cprezzi/grbl-LPC

cprezzi committed:
Workaround for coolant mist control

cprezzi committed:
Update README.md

cprezzi committed:
Update README.md

cprezzi committed:
Update README.md

Commit:
Add open drain configuration per axis

Based on code provided by Jim Fong for Cohesion3D

cprezzi committed:
Merge pull request #37 from johnsonm/open-drain

Add open drain configuration per axis

cprezzi committed:
Update README.md

cprezzi committed:
Config changes to compile for C3d Mini / K40

cprezzi committed:
Merge branch 'master' of https://github.com/cprezzi/grbl-LPC

cprezzi committed:
Compiled bin files for all boards

Commit:
Fix for safety door, feed hold and cycle start bits

uint8_t overflows the 32 bit status register

cprezzi committed:
Merge pull request #42 from betgear/door_fix

Fix for safety door, feed hold and cycle start bits.

tscofield committed:
Fix for issue #49

Correct 32bit mask for probe invert.

Probe Physically Open
$6=0
<Alarm|WPos:0.000,0.000,0.000,0.000|FS:0,0|Pn:ZA|WCO:0.000,0.000,0.000,0.000>
Probe is not triggered

$6=1
<Alarm|WPos:0.000,0.000,0.000,0.000|FS:0,0|Pn:PZA|Ov:100,100,100>
Probe is triggered

Probe Physically Closed
$6=0
<Alarm|WPos:0.000,0.000,0.000,0.000|FS:0,0|Pn:ZA|WCO:0.000,0.000,0.000,0.000>
Probe is not triggered

$6=1
<Alarm|WPos:0.000,0.000,0.000,0.000|FS:0,0|Pn:PZA|Ov:100,100,100>
Probe is triggered

cprezzi committed:
Merge pull request #53 from tscofield/probe

Fix for issue #49 (probe invert mask 8 bit instead of 32)

cprezzi committed:
Update README.md

cprezzi committed:
Update cpu_map.h

Corrected wrong digipot wiper mapping for MKS SBASE boards.

cprezzi committed:
Update cpu_map.h

Correcting back to former wiper registers
